README.md
# Express Prometheus Middleware

[![npm](https://img.shields.io/npm/v/express-prometheus-middleware.svg)](https://www.npmjs.com/package/express-prometheus-middleware)
[![Dependency Status](https://david-dm.org/joao-fontenele/express-prometheus-middleware.svg)](https://david-dm.org/joao-fontenele/express-prometheus-middleware)
[![devDependency Status](https://david-dm.org/joao-fontenele/express-prometheus-middleware/dev-status.svg)](https://david-dm.org/joao-fontenele/express-prometheus-middleware#info=devDependencies)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)
[![Gitter chat](https://badges.gitter.im/gitterHQ/gitter.png)](https://gitter.im/express-prometheus-middleware/community)

This is a middleware for express servers, that expose metrics for prometheus.

The metrics exposed allows to calculate common RED (Request, Error rate, Duration of requests), and USE (Utilisation, Error rate, and Saturation), metrics

## Install

``` bash
yarn add express-prometheus-middleware
# or
npm i --save express-prometheus-middleware
```

## Usage

### Options

| Name | Description | Default |
| :-: | :- | :- |
| metricsPath | Url route that will expose the metrics for scraping. | `/metrics` |
| metricsApp  | Express app that will expose metrics endpoint, if an app is provided, use it, instead of instantiating a new one | `null` |
| collectDefaultMetrics | Whether or not to collect `prom-client` default metrics. These metrics are usefull for collecting saturation metrics, for example. | `true` |
| collectGCMetrics | Whether or not to collect garbage collection metrics via module `prometheus-gc-stats`. Dependency `prometheus-gc-stats` is marked as optional, hence if this option is set to `true` but npm/yarn could not install the dependency, no garbage collection metric will be collected. | `false` |
| requestDurationBuckets | Buckets for the request duration metrics (in seconds) histogram | Uses `prom-client` utility: `Prometheus.exponentialBuckets(0.05, 1.75, 8)` |
| requestLengthBuckets | Buckets for the request length metrics (in bytes) histogram | no buckets (The request length metrics are not collected): `[]` |
| responseLengthBuckets | Buckets for the response length metrics (in bytes) histogram | no buckets (The response length metrics are not collected) `[]` |
| extraMasks | Optional, list of regexes to be used as argument to [url-value-parser](https://www.npmjs.com/package/url-value-parser), this will cause extra route params,  to be replaced with a `#val` placeholder.  | no extra masks: `[]` |
| authenticate | Optional authentication callback, the function should receive as argument, the `req` object and return truthy for sucessfull authentication, or falsy, otherwise. This option supports Promise results. | `null` |
| prefix | Optional prefix for the metrics name | no prefix added |
| customLabels | Optional Array containing extra labels, used together with  `transformLabels` | no extra labels: `[]` |
| transformLabels | Optional `function(labels, req, res)` adds to the labels object dynamic values for each label in `customLabels` | `null` |
| normalizeStatus | Optional parameter to disable normalization of the status code. Example of normalized and non-normalized status code respectively: 4xx and 422.| true
| resetOnScrape | Optional parameter to reset the metrics in the default registry after every scrape| true
### Example

```js
const express = require('express');
const promMid = require('express-prometheus-middleware');
const app = express();

const PORT = 9091;
app.use(promMid({
  metricsPath: '/metrics',
  collectDefaultMetrics: true,
  requestDurationBuckets: [0.1, 0.5, 1, 1.5],
  requestLengthBuckets: [512, 1024, 5120, 10240, 51200, 102400],
  responseLengthBuckets: [512, 1024, 5120, 10240, 51200, 102400],
  /**
   * Uncomenting the `authenticate` callback will make the `metricsPath` route
   * require authentication. This authentication callback can make a simple
   * basic auth test, or even query a remote server to validate access.
   * To access /metrics you could do:
   * curl -X GET user:password@localhost:9091/metrics
   */
  // authenticate: req => req.headers.authorization === 'Basic dXNlcjpwYXNzd29yZA==',
  /**
   * Uncommenting the `extraMasks` config will use the list of regexes to
   * reformat URL path names and replace the values found with a placeholder value
  */
  // extraMasks: [/..:..:..:..:..:../],
  /**
   * The prefix option will cause all metrics to have the given prefix.
   * E.g.: `app_prefix_http_requests_total`
   */
  // prefix: 'app_prefix_',
  /**
   * Can add custom labels with customLabels and transformLabels options
   */
  // customLabels: ['contentType'],
  // transformLabels(labels, req) {
  //   // eslint-disable-next-line no-param-reassign
  //   labels.contentType = req.headers['content-type'];
  // },
}));

// curl -X GET localhost:9091/hello?name=Chuck%20Norris
app.get('/hello', (req, res) => {
  console.log('GET /hello');
  const { name = 'Anon' } = req.query;
  res.json({ message: `Hello, ${name}!` });
});

app.listen(PORT, () => {
  console.log(`Example api is listening on http://localhost:${PORT}`);
});
```

### Metrics exposed

- Default metrics from [prom-client](https://github.com/siimon/prom-client)
- `http_requests_total`: Counter for total requests received, has labels `route`, `method`, `status`
- `http_request_duration_seconds`: - Duration of HTTP requests in seconds, has labels `route`, `method`, `status`

The labels `route` and `status` are normalized:

- `route`: will normalize id like route params
- `status`: will normalize to status code family groups, like `2XX` or `4XX`.

### Example prometheus queries

In the examples below, Suppose you tagged your application as "myapp", in the prometheus scrapping config.

#### Running instances

```js
sum(up{app="myapp"})
```

#### Overall error rate

Rate of http status code 5XX responses

```js
sum(rate(http_requests_total{status="5XX", app="myapp"}[5m]))
```

#### 95% of requests served within seconds

```js
histogram_quantile(0.95, sum(rate(http_request_duration_seconds_bucket{app="myapp"}[5m])) by (le))
```

#### 95% of request length

```js
histogram_quantile(0.95, sum(rate(http_request_length_bytes_bucket{app="myapp"}[5m])) by (le))
```

#### 95% of response length

```js
histogram_quantile(0.95, sum(rate(http_response_length_bytes_bucket{app="myapp"}[5m])) by (le))
```

#### Average response time in seconds

```js
sum(rate(http_request_duration_seconds_sum{app="myapp"}[5m])) by (instance) / sum(rate(http_request_duration_seconds_count{app="myapp"}[5m])) by (instance)
```

#### Overall Request rate

```js
sum(rate(http_requests_total{app="myapp"}[5m])) by (instance)
```

#### Request rate by route

In this example we are removing some health/status-check routes, replace them with your needs.

```js
sum(rate(http_requests_total{app="myapp", route!~"/|/healthz"}[5m])) by (instance, route)
```

#### CPU usage

```js
rate(process_cpu_system_seconds_total{app="myapp"}[5m])
rate(process_cpu_user_seconds_total{app="myapp"}[5m])
```

#### Memory usage

```js
nodejs_heap_size_total_bytes{app="myapp"}
nodejs_heap_size_used_bytes{app="myapp"}
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.2.0](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v1.1.0...v1.2.0) (2021-04-24)


### Features

* add metrics for request length and response length ([30ec2ed](https://github.com/joao-fontenele/express-prometheus-middleware/commit/30ec2eddf660858a22a6677571d8f3afd022d241))


### Bug Fixes

* commitlint misreporting release commit ([cbd4101](https://github.com/joao-fontenele/express-prometheus-middleware/commit/cbd410192dd7bf8003a56471ecae145274377267))

## [1.1.0](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v1.0.0...v1.1.0) (2021-02-19)


### Features

* alter to normalization of status code to an function optional ([09cfa46](https://github.com/joao-fontenele/express-prometheus-middleware/commit/09cfa46b9cfc6c61260f4cb3e7a9b59a81b8e489))
* extend compatibility with prom-client v13.x ([e18732b](https://github.com/joao-fontenele/express-prometheus-middleware/commit/e18732bc11fef7474b7dd0bad50b52982ea11edc))

## [1.0.0](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.9.6...v1.0.0) (2020-10-26)


### ⚠ BREAKING CHANGES

* The /metrics route will not be exposed through the middleware if the 'metricsApp' option is provided.

### Features

* allow metrics route to be exposed on separate port ([69af583](https://github.com/joao-fontenele/express-prometheus-middleware/commit/69af583b24d40bbec5bfd1181d806832052669b1))

### [0.9.6](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.9.5...v0.9.6) (2020-05-18)

### [0.9.5](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.9.4...v0.9.5) (2020-05-15)


### Features

* add posibility to add custom labels ([7e48758](https://github.com/joao-fontenele/express-prometheus-middleware/commit/7e487584d8ebe073de4f87bb478eba1e8cdd2205))

### [0.9.4](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.9.3...v0.9.4) (2020-05-13)


### Features

* add optional dependency to collect garbage collector metrics too ([6a1eb87](https://github.com/joao-fontenele/express-prometheus-middleware/commit/6a1eb875c097269661a0d2c36530b6188bb256ae))

### [0.9.3](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.9.2...v0.9.3) (2020-05-13)

### [0.9.2](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.9.1...v0.9.2) (2020-05-13)

### [0.9.1](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.9.0...v0.9.1) (2020-05-06)

## [0.9.0](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.8.5...v0.9.0) (2020-05-06)


### ⚠ BREAKING CHANGES

* to migrate, install express and prom-client as dependencies in
your own project

### Bug Fixes

* use prom-client and express as peer dependencies ([4825c8c](https://github.com/joao-fontenele/express-prometheus-middleware/commit/4825c8c409f391d0797e5c02ddb03868c108db6e)), closes [#22](https://github.com/joao-fontenele/express-prometheus-middleware/issues/22)

### [0.8.5](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.8.4...v0.8.5) (2019-11-23)


### Features

* **prefix:** Allow adding metrics prefix name ([ead5c1a](https://github.com/joao-fontenele/express-prometheus-middleware/commit/ead5c1a4ecf1ba2888784050589c4ed359b8f182))

### [0.8.4](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.8.3...v0.8.4) (2019-11-15)


### Features

* Allow specifying extraMasks for URL normalizer ([428a767](https://github.com/joao-fontenele/express-prometheus-middleware/commit/428a767048c6405347595c48b73e73efcfcad817))

### [0.8.3](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.8.2...v0.8.3) (2019-11-10)

<a name="0.8.2"></a>
## [0.8.2](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.8.1...v0.8.2) (2019-11-10)



<a name="0.8.1"></a>
## [0.8.1](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.8.0...v0.8.1) (2019-11-10)



<a name="0.8.0"></a>
# [0.8.0](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.7.0...v0.8.0) (2019-11-10)


### Features

* allow metrics to be served under a custom app ([06bbdba](https://github.com/joao-fontenele/express-prometheus-middleware/commit/06bbdba))



<a name="0.7.0"></a>
# [0.7.0](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.6.3...v0.7.0) (2019-10-23)


### Features

* add an authetication option callback to the metrics route ([4df5fe9](https://github.com/joao-fontenele/express-prometheus-middleware/commit/4df5fe9))



<a name="0.6.3"></a>
## [0.6.3](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.6.2...v0.6.3) (2019-10-20)



<a name="0.6.2"></a>
## [0.6.2](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.6.1...v0.6.2) (2019-10-07)


### Bug Fixes

* Removes `x-powered-by` header ([f13a3a5](https://github.com/joao-fontenele/express-prometheus-middleware/commit/f13a3a5)), closes [#3](https://github.com/joao-fontenele/express-prometheus-middleware/issues/3)



<a name="0.6.1"></a>
## [0.6.1](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.6.0...v0.6.1) (2018-08-26)



<a name="0.6.0"></a>
# [0.6.0](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.5.1...v0.6.0) (2018-08-24)


### Features

* change http duration metric from milliseconds to seconds ([fda644c](https://github.com/joao-fontenele/express-prometheus-middleware/commit/fda644c))



<a name="0.5.1"></a>
## [0.5.1](https://github.com/joao-fontenele/express-prometheus-middleware/compare/v0.5.0...v0.5.1) (2018-08-23)


### Bug Fixes


src/metrics.js
const Prometheus = require('prom-client');

/**
 * @param prefix - metrics name prefix
 * request counter
 */
function requestCountGenerator(labelNames, prefix = '') {
  return new Prometheus.Counter({
    name: `${prefix}http_requests_total`,
    help: 'Counter for total requests received',
    labelNames,
  });
}

/**
 * @param {!Array} buckets - array of numbers, representing the buckets for
 * @param prefix - metrics name prefix
 * request duration
 */
function requestDurationGenerator(labelNames, buckets, prefix = '') {
  return new Prometheus.Histogram({
    name: `${prefix}http_request_duration_seconds`,
    help: 'Duration of HTTP requests in seconds',
    labelNames,
    buckets,
  });
}

/**
 * @param {!Array} buckets - array of numbers, representing the buckets for
 * @param prefix - metrics name prefix
 * request length
 */
function requestLengthGenerator(labelNames, buckets, prefix = '') {
  return new Prometheus.Histogram({
    name: `${prefix}http_request_length_bytes`,
    help: 'Content-Length of HTTP request',
    labelNames,
    buckets,
  });
}

/**
 * @param {!Array} buckets - array of numbers, representing the buckets for
 * @param prefix - metrics name prefix
 * response length
 */
function responseLengthGenerator(labelNames, buckets, prefix = '') {
  return new Prometheus.Histogram({
    name: `${prefix}http_response_length_bytes`,
    help: 'Content-Length of HTTP response',
    labelNames,
    buckets,
  });
}

module.exports = {
  requestCountGenerator,
  requestDurationGenerator,
  requestLengthGenerator,
  responseLengthGenerator,
};


src/normalizers.js
const url = require('url');
const UrlValueParser = require('url-value-parser');

const urlParser = new UrlValueParser();

/**
 * Normalizes urls paths.
 *
 * This function replaces route params like ids, with a placeholder, so we can
 * set the metrics label, correctly. E.g., both routes
 *
 * - /api/v1/user/1
 * - /api/v1/user/2
 *
 * represents the same logical route, and we want to group them together,
 * hence the need for the normalization.
 *
 * @param {!string} path - url path.
 * @param {string} [placeholder='#val'] - the placeholder that will replace id
 * like params in the url path.
 * @returns {string} a normalized path, withoud ids.
 */
function normalizePath(originalUrl, placeholder = '#val') {
  const { pathname } = url.parse(originalUrl);
  return urlParser.replacePathValues(pathname, placeholder);
}

/**
 * Normalizes http status codes.
 *
 * Returns strings in the format (2|3|4|5)XX.
 *
 * @param {!number} status - status code of the requests
 * @returns {string} the normalized status code.
 */
function normalizeStatusCode(status) {
  if (status >= 200 && status < 300) {
    return '2XX';
  }

  if (status >= 300 && status < 400) {
    return '3XX';
  }

  if (status >= 400 && status < 500) {
    return '4XX';
  }

  return '5XX';
}

module.exports = {
  normalizePath,
  normalizeStatusCode,
};


src/index.js
const express = require('express');
const Prometheus = require('prom-client');
const ResponseTime = require('response-time');

const {
  requestCountGenerator,
  requestDurationGenerator,
  requestLengthGenerator,
  responseLengthGenerator,
} = require('./metrics');

const {
  normalizeStatusCode,
  normalizePath,
} = require('./normalizers');

const defaultOptions = {
  metricsPath: '/metrics',
  metricsApp: null,
  authenticate: null,
  collectDefaultMetrics: true,
  collectGCMetrics: false,
  // buckets for response time from 0.05s to 2.5s
  // these are arbitrary values since i dont know any better ¯\_(ツ)_/¯
  requestDurationBuckets: Prometheus.exponentialBuckets(0.05, 1.75, 8),
  requestLengthBuckets: [],
  responseLengthBuckets: [],
  extraMasks: [],
  customLabels: [],
  transformLabels: null,
  normalizeStatus: true,
  resetOnScrape: false,
};

module.exports = (userOptions = {}) => {
  const options = { ...defaultOptions, ...userOptions };
  const originalLabels = ['route', 'method', 'status'];
  options.customLabels = new Set([...originalLabels, ...options.customLabels]);
  options.customLabels = [...options.customLabels];
  const { metricsPath, metricsApp, normalizeStatus } = options;

  const app = express();
  app.disable('x-powered-by');

  const requestDuration = requestDurationGenerator(
    options.customLabels,
    options.requestDurationBuckets,
    options.prefix,
  );
  const requestCount = requestCountGenerator(
    options.customLabels,
    options.prefix,
  );
  const requestLength = requestLengthGenerator(
    options.customLabels,
    options.requestLengthBuckets,
    options.prefix,
  );
  const responseLength = responseLengthGenerator(
    options.customLabels,
    options.responseLengthBuckets,
    options.prefix,
  );

  /**
   * Corresponds to the R(equest rate), E(error rate), and D(uration of requests),
   * of the RED metrics.
   */
  const redMiddleware = ResponseTime((req, res, time) => {
    const { originalUrl, method } = req;
    // will replace ids from the route with `#val` placeholder this serves to
    // measure the same routes, e.g., /image/id1, and /image/id2, will be
    // treated as the same route
    const route = normalizePath(originalUrl, options.extraMasks);

    if (route !== metricsPath) {
      const status = normalizeStatus
        ? normalizeStatusCode(res.statusCode) : res.statusCode.toString();

      const labels = { route, method, status };

      if (typeof options.transformLabels === 'function') {
        options.transformLabels(labels, req, res);
      }
      requestCount.inc(labels);

      // observe normalizing to seconds
      requestDuration.observe(labels, time / 1000);

      // observe request length
      if (options.requestLengthBuckets.length) {
        const reqLength = req.get('Content-Length');
        if (reqLength) {
          requestLength.observe(labels, Number(reqLength));
        }
      }

      // observe response length
      if (options.responseLengthBuckets.length) {
        const resLength = res.get('Content-Length');
        if (resLength) {
          responseLength.observe(labels, Number(resLength));
        }
      }
    }
  });

  if (options.collectDefaultMetrics) {
    // when this file is required, we will start to collect automatically
    // default metrics include common cpu and head usage metrics that can be
    // used to calculate saturation of the service
    Prometheus.collectDefaultMetrics({
      prefix: options.prefix,
    });
  }

  if (options.collectGCMetrics) {
    // if the option has been turned on, we start collecting garbage
    // collector metrics too. using try/catch because the dependency is
    // optional and it could not be installed
    try {
      /* eslint-disable global-require */
      /* eslint-disable import/no-extraneous-dependencies */
      const gcStats = require('prometheus-gc-stats');
      /* eslint-enable import/no-extraneous-dependencies */
      /* eslint-enable global-require */
      const startGcStats = gcStats(Prometheus.register, {
        prefix: options.prefix,
      });
      startGcStats();
    } catch (err) {
      // the dependency has not been installed, skipping
    }
  }

  app.use(redMiddleware);

  /**
   * Metrics route to be used by prometheus to scrape metrics
   */
  const routeApp = metricsApp || app;
  routeApp.get(metricsPath, async (req, res, next) => {
    if (typeof options.authenticate === 'function') {
      let result = null;
      try {
        result = await options.authenticate(req);
      } catch (err) {
        // treat errors as failures to authenticate
      }

      // the requester failed to authenticate, then return next, so we don't
      // hint at the existance of this route
      if (!result) {
        return next();
      }
    }

    res.set('Content-Type', Prometheus.register.contentType);
    const result = res.end(await Prometheus.register.metrics());
    if (options.resetOnScrape) {
      Prometheus.register.resetMetrics();
    }
    return result;
  });

  return app;
};
